### ISSUE TITLE
MKLDNN convolution leaks memory when input sizes vary

### ISSUE THREAD

**Issue by FerranPares** [pytorch/pytorch#32037]
## 🐛 Bug

Cpu memory leak when using the `collate_fn` of the DataLoader class.

## To Reproduce

Run the following code to reproduce the memory leak:
```python
import os
import subprocess
import numpy as np
import gc
import torch
import torch.nn as nn
from memory_profiler import profile
from torch.utils.data import Dataset, DataLoader


def current_memory_usage():
    '''Returns current memory usage (in MB) of a current process'''
    out = subprocess.Popen(['ps', '-p', str(os.getpid()), '-o', 'rss'],
                           stdout=subprocess.PIPE).communicate()[0].split(b'\n')
    mem = float(out[1].strip()) / 1024
    return mem


class RandomDataset(Dataset):

    def __init__(self, n_samples):
        self.n_samples = n_samples

    def __len__(self):
        return self.n_samples

    def __getitem__(self, idx):
        width = np.random.randint(200, 401)
        height = np.random.randint(200, 401)
        image = np.random.rand(3, width, height)
        return image, np.random.randint(0, 2)


def padding_fn(batch):
    images = [item[0] for item in batch]
    labels = [item[1] for item in batch]

    max_width = max(img.shape[1] for img in images)
    max_height = max(img.shape[2] for img in images)

    padded_images = []
    for img in images:
        right_pad = max_width - img.shape[1]
        bottom_pad = max_height - img.shape[2]
        padding_tuple = ((0, 0), (0, right_pad), (0, bottom_pad))
        padded_img = np.pad(img, padding_tuple)
        padded_images.append(torch.from_numpy(padded_img.astype(np.float32)))

    images_batch = torch.stack(padded_images)
    labels_batch = torch.stack([torch.from_numpy(np.array(lab)) for lab in labels])
    return images_batch, labels_batch


class BasicModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 64, kernel_size=3)
        self.relu1 = nn.ReLU()
        self.conv2 = nn.Conv2d(64, 64, kernel_size=3)
        self.relu2 = nn.ReLU()

    def forward(self, x):
        x = self.conv1(x)
        x = self.relu1(x)
        x = self.conv2(x)
        x = self.relu2(x)
        return x


@profile
def main():
    random_ds = RandomDataset(n_samples=100)
    input_pipe = DataLoader(random_ds, batch_size=2, collate_fn=padding_fn)
    model = BasicModel()

    for epoch in range(1000):
        for i_step, (batch_images, batch_labels) in enumerate(input_pipe):
            model(batch_images)

            print('{}-{}:  \t\tMemory usage: {:.1f}MB'.format(epoch, i_step, current_memory_usage()))
            gc.collect()


if __name__ == "__main__":
    main()

```

The output of this code produces a constant increase of memory until the process gets killed:

> 0-0:  		Memory usage: 205.3MB
> 0-1:  		Memory usage: 259.4MB
> 0-2:  		Memory usage: 321.4MB
> 0-3:  		Memory usage: 377.4MB
> 0-4:  		Memory usage: 427.1MB
> 0-5:  		Memory usage: 457.7MB
> 0-6:  		Memory usage: 523.6MB
> 0-7:  		Memory usage: 590.3MB
> 0-8:  		Memory usage: 647.0MB
> 0-9:  		Memory usage: 693.1MB
> 0-10:  		Memory usage: 753.0MB
> 0-11:  		Memory usage: 814.9MB
> 0-12:  		Memory usage: 867.7MB
> 0-13:  		Memory usage: 940.9MB
> 0-14:  		Memory usage: 984.4MB
> 0-15:  		Memory usage: 1052.4MB
> 0-16:  		Memory usage: 1113.0MB
> 0-17:  		Memory usage: 1147.2MB
> 0-18:  		Memory usage: 1214.5MB
> 0-19:  		Memory usage: 1256.2MB
> 0-20:  		Memory usage: 1321.4MB
> 0-21:  		Memory usage: 1363.3MB
> 0-22:  		Memory usage: 1398.8MB
> 0-23:  		Memory usage: 1461.3MB
> 0-24:  		Memory usage: 1526.1MB
> 0-25:  		Memory usage: 1589.6MB
> 0-26:  		Memory usage: 1654.9MB
> 0-27:  		Memory usage: 1714.5MB
> 0-28:  		Memory usage: 1765.7MB
> 0-29:  		Memory usage: 1822.8MB
> 0-30:  		Memory usage: 1867.4MB
> 0-31:  		Memory usage: 1930.6MB
> 0-32:  		Memory usage: 2003.0MB
> 0-33:  		Memory usage: 2052.4MB
> 0-34:  		Memory usage: 2101.6MB
> 0-35:  		Memory usage: 2146.0MB
> 0-36:  		Memory usage: 2202.7MB
> 0-37:  		Memory usage: 2260.3MB
> 0-38:  		Memory usage: 2328.6MB
> 0-39:  		Memory usage: 2382.4MB
> 0-40:  		Memory usage: 2451.6MB
> 0-41:  		Memory usage: 2599.9MB
> 0-42:  		Memory usage: 2652.5MB
> 0-43:  		Memory usage: 2606.8MB
> 0-44:  		Memory usage: 2673.3MB
> 0-45:  		Memory usage: 2742.2MB
> 0-46:  		Memory usage: 2792.8MB
> 0-47:  		Memory usage: 2863.5MB
> 0-48:  		Memory usage: 2923.5MB
> 0-49:  		Memory usage: 2989.3MB
> 1-0:  		Memory usage: 3039.4MB
> 1-1:  		Memory usage: 3093.0MB
> 1-2:  		Memory usage: 3152.9MB
> 1-3:  		Memory usage: 3198.4MB
> 1-4:  		Memory usage: 3252.2MB
> 1-5:  		Memory usage: 3296.8MB
> 1-6:  		Memory usage: 3348.2MB
> 1-7:  		Memory usage: 3383.5MB
> 1-8:  		Memory usage: 3436.1MB
> 1-9:  		Memory usage: 3511.5MB
> 1-10:  		Memory usage: 3559.3MB
> 1-11:  		Memory usage: 3626.3MB
> 1-12:  		Memory usage: 3678.6MB
> 1-13:  		Memory usage: 3743.9MB
> 1-14:  		Memory usage: 3790.7MB
> 1-15:  		Memory usage: 3845.6MB
> 1-16:  		Memory usage: 3893.4MB
> 1-17:  		Memory usage: 4009.8MB
> 1-18:  		Memory usage: 4074.2MB
> 1-19:  		Memory usage: 4110.5MB
> 1-20:  		Memory usage: 4167.3MB
> 1-21:  		Memory usage: 4219.2MB
> 1-22:  		Memory usage: 4271.2MB
> 1-23:  		Memory usage: 4321.2MB
> 1-24:  		Memory usage: 4384.1MB
> 1-25:  		Memory usage: 4441.2MB
> 1-26:  		Memory usage: 4503.8MB
> 1-27:  		Memory usage: 4572.8MB
> 1-28:  		Memory usage: 4632.8MB
> 1-29:  		Memory usage: 4671.7MB
> 1-30:  		Memory usage: 4718.3MB
> 1-31:  		Memory usage: 4767.4MB
> 1-32:  		Memory usage: 4804.5MB
> 1-33:  		Memory usage: 4873.8MB
> 1-34:  		Memory usage: 4929.9MB
> 1-35:  		Memory usage: 4973.0MB
> 1-36:  		Memory usage: 5035.8MB
> 1-37:  		Memory usage: 5080.9MB
> 1-38:  		Memory usage: 5151.5MB
> 1-39:  		Memory usage: 5207.2MB
> 1-40:  		Memory usage: 5258.6MB
> 1-41:  		Memory usage: 5315.5MB
> 1-42:  		Memory usage: 5370.0MB
> 1-43:  		Memory usage: 5413.9MB
> 1-44:  		Memory usage: 5468.8MB
> 1-45:  		Memory usage: 5529.9MB
> 1-46:  		Memory usage: 5596.8MB
> 1-47:  		Memory usage: 5668.1MB
> 1-48:  		Memory usage: 5713.1MB
> 1-49:  		Memory usage: 5781.3MB
> 2-0:  		Memory usage: 5847.2MB
> 2-1:  		Memory usage: 5884.3MB
> 2-2:  		Memory usage: 5942.9MB
> 2-3:  		Memory usage: 5997.2MB
> 2-4:  		Memory usage: 6098.8MB
> 2-5:  		Memory usage: 6059.6MB
> 2-6:  		Memory usage: 6103.1MB
> 2-7:  		Memory usage: 6164.3MB
> 2-8:  		Memory usage: 6236.9MB
> 2-9:  		Memory usage: 6289.6MB
> 2-10:  		Memory usage: 6335.0MB
> 2-11:  		Memory usage: 6392.7MB
> 2-12:  		Memory usage: 6539.9MB
> 2-13:  		Memory usage: 6514.1MB
> 2-14:  		Memory usage: 6562.0MB
> 2-15:  		Memory usage: 6609.6MB
> 2-16:  		Memory usage: 6616.8MB
> 2-17:  		Memory usage: 6654.4MB
> 2-18:  		Memory usage: 6699.7MB
> 2-19:  		Memory usage: 6749.9MB
> Killed

## Expected behavior

I expected the code above to have constant usage of memory from step to step computation.

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: openSUSE Tumbleweed
GCC version: (SUSE Linux) 9.2.1 20191209 [gcc-9-branch revision 279114]
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.3.1
[conda] Could not collect

## Additional context

I'm working on a CNN training using images of variable shape, so I have to add some padding to some images in the same batch to agree on their shape. For this purpose, I use the `collate_fn` from the `DataLoader` class. This how I end up finding the memory leak issue.

The code contains:
 -  A utility to track the memory usage of the process `current_memory_usage`
 -  A PyTorch dataset `RandomDataset` that generates random images of different shapes ranging width and height between 200 and 400 pixels
 -  A `padding_fn` in charge of padding images that belong to the same batch to ensure that they have the same shape (this function is the one provided to `collate_fn`)
 -  A basic convolutional model `BasicModel`
 -  The main code that iterates over the dataset in batches of 2 images and forwards through a basic CNN.

I can not advance on my project without solving this issue, so this is critical.

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh

**Comment by FerranPares** [comment]
I could avoid the memory leak when substituting the `padding_fn` with a code that returns random tensors (completely ignoring the argument `batch`):
```python
def padding_fn(batch):
    images_batch = torch.rand((2, 3, 100, 100))
    labels_batch = torch.randint(0, 2, (2,))
    return images_batch, labels_batch
```
I'm not familiar with MKLDNN but hope this information helps to find the source of the memory leak.

PD: Sorry for the misleading title.

**Comment by SsnL** [comment]
no worries. a more minimal repro is

```py
import os
import subprocess
import numpy as np
import gc
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader


def current_memory_usage():
    '''Returns current memory usage (in MB) of a current process'''
    out = subprocess.Popen(['ps', '-p', str(os.getpid()), '-o', 'rss'],
                           stdout=subprocess.PIPE).communicate()[0].split(b'\n')
    mem = float(out[1].strip()) / 1024
    return mem


class BasicModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 64, kernel_size=3)
        self.relu1 = nn.ReLU()
        self.conv2 = nn.Conv2d(64, 64, kernel_size=3)
        self.relu2 = nn.ReLU()

    def forward(self, x):
        x = self.conv1(x)
        x = self.relu1(x)
        x = self.conv2(x)
        x = self.relu2(x)
        return x


def main():
    model = BasicModel()

    for epoch in range(1000):
        for i_step in range(200):
            import random 
            batch_images = torch.zeros(2, 3, random.randint(250, 400), random.randint(250, 400))
            model(batch_images)

            print('{}-{}:  \t\tMemory usage: {:.1f}MB'.format(epoch, i_step, current_memory_usage()))
            gc.collect()


if __name__ == "__main__":
    main()
```

**Comment by FerranPares** [comment]
I just tried to avoid the memory leak by disabling MKLDNN:
```python
from torch.utils import mkldnn
mkldnn.enabled = False
```
Unfortunately, I'm not able to avoid it, neither using the original code that I posted, neither using the latest reproduction code you posted. How did you exactly disabled MKLDNN? Am I doing something wrong?

**Comment by vpirogov** [comment]
@XiaobingSuper, @gujinghui could you please help investigate?

**Comment by SsnL** [comment]
@FerranPares Try 
```py
import torch.backends.mkldnn

torch.backends.mkldnn.enabled = False
```

**Comment by XiaobingSuper** [comment]
@SsnL @FerranPares , it seems a same problem said in https://github.com/pytorch/pytorch/issues/27971 which IDEEP have primitive cache for the performance, you can set  **LRU_CACHE_CAPACITY** to a smaller number(the default number is 1024, you can set it to 1) to reduce the memory use. Thanks!

**Comment by rgommers** [comment]
I cannot reproduce this issue with pytorch 1.3 or 1.4, but it's very likely fixed in master, see https://github.com/pytorch/pytorch/issues/27971#issuecomment-629698921.

**Comment by pinzhenx** [comment]
Closing due to lack of activity. Feel free to reopen with new data.
